fog/ingest/server/src/controller.rs

```diff
@@ -379,36 +379,39 @@ where
             // if we are still needed to be active.
             if state.is_active() {
                 log::trace!(self.logger, "publish report");
-                match self.publish_report(&ingress_pubkey, &mut state) {
-                    Ok(ingress_key_status) => {
-                        // If our key is retired, and the index we want to scan is past expiry,
-                        // early return. Note, we don't even NEED to scan
-                        // when block.index == pubkey_expiry, because the
-                        // semantic of pubkey expiry is that it bounds the tombstone block, and a
-                        // transaction cannot land in its tombstone block.
-                        // But scanning the tombstone block as well may help
-                        // deal with off-by-one errors somewhere else, and doesn't really hurt.
-                        if ingress_key_status.retired
-                            && block.index > ingress_key_status.pubkey_expiry
-                        {
-                            log::warn!(self.logger, "When preparing to process block index {}, we discovered that our ingress key is expired: {:?}. Switching to idle and nuking keys.", block.index, ingress_key_status);
-                            state.set_idle();
-                            self.new_egress_key(&mut state)
-                                .expect("Failure to rotate egress key can't be recovered from");
-                            return;
+                let mut retry_seconds = 1;
+                loop {
```

Comment:
Might it make sense to make this a conditional loop? e.g. `while state.is_active()`

Reply:
so in this loop, we have a lock on the state, and nothing else can change us to inactive while we have that lock, so there would be no reason to do that here

Reply:
Ah OK. Maybe document that we hold a lock on `state`? Ignore if it's apparent to someone who's looked at the `fog` code for a while  🙃